lib/spack/spack/util/windows_registry.py

```diff
@@ -0,0 +1,324 @@
+    the registry that is already open. The system defines predefined keys
+    that are always open. Predefined keys help an application navigate in
+    the registry."""
+
+    HKEY_CLASSES_ROOT = _HKEY_CONSTANT("HKEY_CLASSES_ROOT")
+    HKEY_CURRENT_USER = _HKEY_CONSTANT("HKEY_CURRENT_USER")
+    HKEY_USERS = _HKEY_CONSTANT("HKEY_USERS")
+    HKEY_LOCAL_MACHINE = _HKEY_CONSTANT("HKEY_LOCAL_MACHINE")
+    HKEY_CURRENT_CONFIG = _HKEY_CONSTANT("HKEY_CURRENT_CONFIG")
+    HKEY_PERFORMANCE_DATA = _HKEY_CONSTANT("HKEY_PERFORMANCE_DATA")
+
+
+class WindowsRegistryView(object):
+    """
+    Interface to provide access, querying, and searching to Windows registry entries.
+    This class represents a single key entrypoint into the Windows registry
+    and provides an interface to this key's values, its subkeys, and those subkey's values.
+    This class cannot be used to move freely about the registry, only subkeys/values of
+    the root key used to instantiate this class.
+    """
+
+    def __init__(self, key, root_key=HKEY.HKEY_CURRENT_USER):
+        """Constructs a Windows Registry entrypoint to key provided
+        root_key should be an already open root key or an hkey constant if provided
+
+        Args:
+            key (str): registry key to provide root for registry key for this clas
+            root_key: Already open registry key or HKEY constant to provide access into
+                         the Windows registry. Registry access requires an already open key
+                         to get an entrypoint, the HKEY constants are always open, or an already
+                         open key can be used instead.
+        """
+        if not is_windows:
+            raise RuntimeError(
+                "Cannot instantiate Windows Registry class on non Windows platforms"
+            )
+        self.key = key
+        self.root = root_key
+        self._reg = None
+
+    @contextmanager
+    def invalid_reg_ref_error_handler(self):
+        try:
+            try:
+                yield
+            except FileNotFoundError as e:
+                if e.winerror == 2:
+                    tty.debug("Key %s at position %s does not exist" % (self.key, str(self.root)))
+                else:
+                    raise e
+        finally:
+            return
+
+    def __bool__(self):
+        return self.reg != -1
+
+    def _load_key(self):
+        try:
+            self._reg = RegistryKey(
+                os.path.join(str(self.root), self.key),
+                winreg.OpenKeyEx(self.root.hkey, self.key, access=winreg.KEY_READ),
+            )
+        except FileNotFoundError as e:
+            if e.winerror == 2:
+                self._reg = -1
+                tty.debug("Key %s at position %s does not exist" % (self.key, str(self.root)))
+            else:
+                raise e
+
+    def _valid_reg_check(self):
+        if self.reg == -1:
+            tty.debug("Cannot perform operation for nonexistent key %s" % self.key)
```

Comment:
The only way to check if a key exists or not is to query it and allow the query to fail if the key is missing. From that point it becomes difficult to distinguish between misspelling and invalid queries, so I've left it up to the user how they want to handle a missing key but just returning Null. Raising an exception is effectively the same thing, except instead of checking for null a user would need to do some exception handling. If you prefer the exception handling approach, I can remove this and the `invalid_reg_ref_handler` context manager.

Reply:
Based on further conversation examples of missing keys are things like:

* If you query for a key related to an existing msmpi installation and it has not in fact been installed, then it would return nothing
* Same for WDK/SDK queries added by this PR to `detection/common.py`

However, most of the methods that use this shouldn't in fact be run (i.e. `get_subkeys` shouldn't be run on a failed query key) so IMO this should generally raise an exception (the initial query for the root key should be allowed to "miss" though without an exception).